### ISSUE TITLE
Public API suggestion: Make bool RuntimeHelpers.IsKnownConstant(...) public for all primitive types

### ISSUE THREAD

**Issue by gerhard17** [dotnet/runtime#67285]
### Background and motivation

Recently two internal IsKnownConstant() overloads of RuntimeHelpers for char and string? where added.
For a library developer it would be usefull to consume this API like the framework internaly does.

Same performance/codegen reasons apply as with the original (internal) proposal. #11484

At least one overload for each primitive type (+ the existing string) would be wellcomed.

See also: #63734 and #64809

This request has similarities with GNU GCC int __builtin_constant_p (exp)

### API Proposal

```C#
namespace System.Runtime.CompilerServices
{
    public static partial class RuntimeHelpers
    {
        // The following intrinsics return true if input is a compile-time constant
        // Feel free to add more overloads on demand

       // CHANGE VISIBILITY

        [Intrinsic]
        public static bool IsKnownConstant(string? t) => false;

        [Intrinsic]
        public static bool IsKnownConstant(char t) => false;

        // ADD OVERLOADS

        [Intrinsic]
        public static bool IsKnownConstant(byte t) => false;

        [Intrinsic]
        public static bool IsKnownConstant(sbyte t) => false;

        [Intrinsic]
        public static bool IsKnownConstant(short t) => false;

        [Intrinsic]
        public static bool IsKnownConstant(ushort t) => false;

        [Intrinsic]
        public static bool IsKnownConstant(int t) => false;

        [Intrinsic]
        public static bool IsKnownConstant(uint t) => false;

        [Intrinsic]
        public static bool IsKnownConstant(long t) => false;

        [Intrinsic]
        public static bool IsKnownConstant(ulong t) => false;

        [Intrinsic]
        public static bool IsKnownConstant(nint t) => false;

        [Intrinsic]
        public static bool IsKnownConstant(nuint t) => false;

        [Intrinsic]
        public static bool IsKnownConstant(float t) => false;

        [Intrinsic]
        public static bool IsKnownConstant(double t) => false;

    }
}
```

### API Usage

API can be public used like the current internal usage inside of the runtime.

### Alternative Designs

Maybe a true generic Version can be used.

```C#
namespace System.Runtime.CompilerServices
{
    public static partial class RuntimeHelpers
    {
        [Intrinsic]
        public static bool IsKnownConstant<T>(T t) => false;
    }
}
```

### Risks

- No breaking change. Not an public UI today.

- Will depend on JIT beyond c# language spec.

**Comment by teo-tsirpanis** [comment]
I'm in favor of making it generic.

**Comment by SupinePandora43** [comment]
> I'm in favor of making it generic.

Will it be able to consume `null` reference types and `default` value types?

**Comment by EgorBo** [comment]
1) ~~The generic version must be guarded with `: unmanaged` constraint because with ref types it might introduce a runtime lookup that will prevent inlining~~ (no longer relevant, jit can do inlining with lookups now)
2) Is ` __builtin_constant` popular in the C/C++ world? I've never seen it in the wild. Do you have a real-world scenario where it'd help you?
3) There are unresolved concerns with this API and JIT's inliner, e.g.:

```csharp
void DoWork(int a)
{
    if (RuntimeHelpers.IsKnownConstant(a) && a == 42)
    {
        // a lot of code
    }
    // a lot of code
}
```
jit (currently) is not able to give it additional inlining boost only when a == 42 (currently it doesn't boost at all)

**Comment by gerhard17** [comment]
Could restriction unmanaged be dangerous with large structs?
Copying struct data, when no JIT is involved.

**Comment by SupinePandora43** [comment]
> Could restriction unmanaged be dangerous with large structs? Copying struct data, when no JIT is involved. Contradicting the performance gain in non JIT environments.

What about `in` and `ref readonly`?

**Comment by buyaa-n** [comment]
Tag @GrabYourPitchforks for triage as he created the [internal API proposal](https://github.com/dotnet/runtime/issues/11484)

**Comment by gerhard17** [comment]
@jkotas 
How does the annotation `[MethodImpl(MethodImplOptions.AggressiveInlining)]` change the inlining behavior?

Will all code be inlined or is only the budget increased to inline more (but maybe not all) code?

**Comment by GrabYourPitchforks** [comment]
> Tag GrabYourPitchforks for triage as he created the internal API proposal

I agree with Jan's assessment at https://github.com/dotnet/runtime/issues/67285#issuecomment-1081889927. This is experimental and very temperamental. We need more evidence that it's widely useful before exposing it publicly. It honestly wouldn't surprise me if the JIT folks somehow come up with some clever scheme which obviates the utility of _IsKnownConstant_ entirely and we end up removing the API as a result.

**Comment by redknightlois** [comment]
This code should have a `IsKnownConstant` guard in the size. Currently we do it by code inspection. https://github.com/ravendb/ravendb/blob/c0bbece37e7b3d3afff2010b6b0e108633fd136e/src/Sparrow/Memory.cs#L472

**Comment by crystal-rgb** [comment]
I see significant value in exposing this method, not to replace the JIT's constant propagation, but to enable constant-dependent algorithm dispatch. 
If a library author can detect an argument is constant they can use built-in math operations to fold parts of their own computations into constants or switch to some specialised caching techniques for memory management. And because its an optimisation technique its low risk too, if the JIT doesn't inline everything and detects it 100% of the time its fine the only side effect it produces is slightly worse performance for that case.

Even if it doesn't end up being used everywhere in code it will **really** benefit some niche situations where its actually needed.
